### ISSUE TITLE
Verify that FxR sends correct search code to Google when user searches from URL bar

### ISSUE THREAD

**Issue by avrignaud** [MozillaReality/FirefoxReality#248]
Hardware: ALL
Steps to Reproduce: NA
Actual Behavior: NA
Expected Behavior: Verify that FxR sends correct search code to Google (as the sole/default search engine in v1.0) such that URL bar searches are sent to Google appropriately and Mozilla gets credit/paid.
Recommendation: Same

**Comment by bluemarvin** [comment]
We aren't currently doing auto complete, but the actual search doesn't currently contain a code:

https://www.google.com/search?q=XXXX

What should we be using?

**Comment by larsbergstrom** [comment]
We'll work with the business folks to get that to you ASAP, thanks! Few internal bits to sort out.

**Comment by nancyhang** [comment]
Blocked by GV issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1451476

**Comment by bluemarvin** [comment]
The linked GV bug is not related. We just need a code to include in the search.

**Comment by bluemarvin** [comment]
Will the code need to be stored in taskcluster secrets and injected at build time or will it be a public code?

**Comment by nancyhang** [comment]
Deleted the comment on GV bug, its on another issue.

**Comment by keianhzo** [comment]
@larsbergstrom @avrignaud any updates on the code to be used?

**Comment by larsbergstrom** [comment]
@keianhzo We're working with legal / etc. right now to figure out some of the details - should have answers in a day or two!

**Comment by keianhzo** [comment]
@larsbergstrom @mkaply Does this apply to all the headsets x stores? There is no way to know what store the app has been downloaded from at runtime so in case stores support multi apk per country we'll need to make different builds for each store/headset to bundle the right code, otherwise I think we'll need to go straight to P2 and check coarse location (This will require the user to grant the location permission). I'll dig into multiple apks in stores and headsets GPS support to see what are the real alternatives.

**Comment by larsbergstrom** [comment]
My understanding would be that if you need the absolute most straightforward thing, the default fallback should be the Google "rest of world" code.

The intent with the store check was not to check where the APK had been downloaded from, but which locale the store is currently set to as the best indicator of location.

A recommendation was *not* to use the system locale, as that is wrong in a very significant number of cases (usually resulting in using the US code elsewhere in the world).

w.r.t. the coarse location, I believe that we (Mozilla) have a geolocation service we can use that powers `navigator.geolocation` on platforms such as desktop that doesn't require using the OS geoloc prompts. @mkapy might know more (otherwise we can bug @jdm <grin>).

**Comment by mkaply** [comment]
yes, we have a geolocation service that can be used:

https://location.services.mozilla.com/v1/country?key=%MOZILLA_API_KEY%

When in doubt for Google, always default to the rest of world code.

As far as the other areas (russia, china, etc), we'll have to figure out a way to set the default engine for those countries.

**Comment by cvan** [comment]
Thanks. Yep, we should be using the server-side geolocation endpoint that @mkaply provided above.

It's used in Firefox's New Tab page:

```js
    // Download the user's country using the geolocation service.
    // Please do not directly use this code or Snippets key.
    // Contact MLS team for your own credentials.
    // https://location.services.mozilla.com/contact
function downloadUserCountry() {
        var GEO_URL = "https://location.services.mozilla.com/v1/country?key=fff72d56-b040-4205-9a11-82feda9d83a3";
        var request = new XMLHttpRequest();
        request.timeout = 60000;
        request.onreadystatechange = function() {
            if (request.readyState == 4 && request.status == 200) {
                var country_data = JSON.parse(request.responseText);

                try {
                    gSnippetsMap.set('geoCountry', country_data.country_code);
                    gSnippetsMap.set('geoLastUpdated', new Date());
                } catch (e) {
                    // Most likely failed to load Data file. Continue on without us,
                    // we'll try again next time.
                }
            }
        };
        request.open('GET', GEO_URL, true);
        request.send();
    }
```

@larsbergstrom: I presume Firefox Reality should have its own API key. Once we have that, geolocation is a `GET` request away.

**Comment by cvan** [comment]
> @larsbergstrom @mkaply Does this apply to all the headsets x countries? There is no way to know what store the app has been downloaded from at runtime so in case stores support multi apk per country we'll need to make different builds for each store/headset to bundle the right code, otherwise I think we'll need to go straight to P2 and check coarse location (This will require the user to grant the location permission). I'll dig into multiple apks/app availablility in stores and headsets GPS support to see what are the real alternatives.

@keianhzo: like how Taskcluster generates separate flavors for each platform (in both debug and release versions) could we also have the build system generate a per-store APK (which would be otherwise exactly the same as that flavor's APK, except it would include a string constant that specified `APP_STORE` for the app store to which it would be submitted)?

**Comment by bluemarvin** [comment]
We will know what platform we are running on which will correlate 1:1 to the app store. However, we won't know where it is running with out additional information. For example the current Wave Viveport app store seems to just throw all regions into one store.

**Comment by cvan** [comment]
Using Mozilla's aforementioned Geolocation API service sounds like the best way to determine the region. When first installed, assume `region` is `Rest of the World`. Make a `GET` request to the API. If connected to the Internet and the API can be reached and geocoded to a known region, update stored `region`. Could add some logic for retrying and updating when network availability changes.

For the _New Tab_ page, I was going to use `window`'s `online` and `offline` events to know when to fetch a JSON endpoint which includes settings, new/updated metadata about featured sites. I was going to proxy the Geolocation API on my server to update the `region`.

But, we could save some network requests (and possible mismatching) by having platform code make the client requests for determining the region. And in #324 I could pass `?region=<region>` (or `postMessage(region, …)`).

**Comment by keianhzo** [comment]
I've checked and although PlayStore supports it, stores like Oculus don't support multi apk per country so there is no way of knowing which store/country the APK it's been downloaded from at runtime. There is also no way of getting that information from the platform store service.

I guess the best approach will be to use the Mozilla geolocation service which is great as we avoid asking for any permissions. 

We can request the location upon app startup and refresh a "region" setting that will be persisted. If the region is any of the special cases we persist it's code otherwise use the rest of the world code. We can add some retry policy in case the initial request fails. I think that's the most simple approach. If required we could make it a background service that refreshes the region periodically every day or X hours and maybe refactor it as an android-component in case it's useful for other products.

**Comment by keianhzo** [comment]
@larsbergstrom @mkaply Do we have a FxR API key for Mozilla geolocation service? Should it be hardcoded or injected via Taskcluster?

**Comment by cvan** [comment]
@keianhzo I talked to Victor Ng on IRC (vng in `#geo`), and he told me 

and he said to do a `GET` on https://location.services.mozilla.com/v1/country without an API key (so can remove the `?key=…`), and we'll be fine without limitations.

btw, very nice work on the [`search_code` branch](https://github.com/MozillaReality/FirefoxReality/compare/search_code) you're working on. that was quick!

(the full conversation if you're curious:)

<details>

```
2:36 PM <cvan> Christopher Van Wiemeersch hi there, curious if anyone can help with an API key for Firefox Reality (our VR/AR browser)
2:36 PM context here: https://github.com/MozillaReality/FirefoxReality/issues/248#issuecomment-412878203
2:37 PM https://github.com/MozillaReality/FirefoxReality/compare/search_code#diff-c80c8b1038c63c52bcbc77d915811941R19
2:37 PM <vng> hey - we've been advised to not hand out keys for internal projects at this time
2:47 PM <cvan> Christopher Van Wiemeersch oh, okay. I assume can't go into the details?
2:47 PM is there a different service we are allowed to use? we're doing a public release to app stores in mid-September
2:52 PM <vng> cvan: can't get into details other than "legal dept told me so".  there might be a way around though
2:52 PM looks like that bug means you only need region specific queries
2:52 PM <vng> for country
2:52 PM https://location.services.mozilla.com/v1/country will do the trick without an API key
2:52 PM alternatively, we have licenses to a maxmind database - but i'm not familiar with our licensing agreement and the details of how we're allowed to use it
2:53 PM but a basic GET on that URL i just sent - without an API key - will do the right thing.  that'll do a simple IP based lookup against maxmind
```

</details>

**Comment by keianhzo** [comment]
Thanks @cvan ! I've updated that. 

I've observed that some engines redirect to the main engine page. ie. yandex.by/search redirects to yandex.ru/search but only on the mobile version (Chrome and Fennec) on Desktop browsers it works as expected.

Also I don't see in https://dxr.mozilla.org/mozilla-central/source/browser/components/search/searchplugins specific code for Google US. Maybe @mkaply can help with that.
